File: go/src/chapter_2/06_palindrome.go
// Реализуйте функцию, проверяющую, является ли связный список палиндромом.
package chapter_2

// Для двусвязного списка
func (ll *LinkedList) IsPalindrome() bool {
	if ll.head == nil || ll.head == ll.tail {
		return true
	}
	h, t := ll.head, ll.tail
	for {
		if h.value != t.value {
			return false
		}
		h = h.next
		t = t.prev
		if h == t || h.prev == t {
			break
		}
	}
	return true
}

// Для односвязного списка с использованием стека
func (ll *LinkedList) IsPalindromeStack() bool {
	left, right := ll.head, ll.head
	stack := []int{}

	for right != nil && right.next != nil {
		stack = append(stack, left.value)
		left = left.next
		right = right.next.next
	} 

	if right != nil {
		left = left.next
	}

	for left != nil {
		if i := pop(&stack); i != left.value {
			return false
		}
		left = left.next
	}

	return true
}

func pop(stack *[]int) int {
	last := len(*stack) - 1
	item := (*stack)[last]
	*stack = (*stack)[:last]
	return item
}

File: python/chapter_2/06_palindrome.py
# Реализуйте функцию, проверяющую, является ли связный список палиндромом.
import unittest
from LinkedList import LinkedList


def is_palindrome(data):
    ll = LinkedList()
    ll.insert_multiple(*data)
    return str(ll) == str(ll.reverse())


def is_palindrome_stack(data):
    ll = LinkedList()
    ll.insert_multiple(*data)

    left = right = ll.start_node
    stack = []
    while right and right.ref:
        stack.append(left.item)
        left = left.ref
        right = right.ref.ref

    if right:
        left = left.ref

    while left:
        if stack.pop() != left.item:
            return False
        left = left.ref

    return True


class Test(unittest.TestCase):
    cases = [
        ([1, 2, 3, 4, 5, 4, 3, 2, 1], True),
        ([1, 2, 3, 4, 5, 6, 3, 2, 1], False)
    ]

    def test_palindrome(self):
        for data, expect in self.cases:
            self.assertEqual(expect, is_palindrome(data))

    def test_palindrome_stack(self):
        for data, expect in self.cases:
            self.assertEqual(expect, is_palindrome_stack(data))


if __name__ == "__main__":
    unittest.main()
